Add throttler to limit the ResourceMonitor activity.

basuke opened the pull request:
#### 69f96c73ab392b6e7969da02996ce934f7951757
<pre>
Add throttler to limit the ResourceMonitor activity.
<a href="https://bugs.webkit.org/show_bug.cgi?id=285107">https://bugs.webkit.org/show_bug.cgi?id=285107</a>
<a href="https://rdar.apple.com/141840417">rdar://141840417</a>

Reviewed by Chris Dumez.

When ResourceMonitor detects large resource usage in a iframe, currently the decision is in
action immediately. In this patch, add throttler to limit the execution of the decision.

- Moving the action from LocalFrame to LocalFrameLoaderClient.
- Ask WebPageProxy for the decision.
- Add ResourceMonitorThrottler to limit the action.
- Place the throttler in ResourceMonitorManager for the global decision.
- Send the result to WebPage if the decision is ongoing.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/loader/LocalFrameLoaderClient.cpp:
(WebCore::LocalFrameLoaderClient::didExceedNetworkUsageThreshold):
* Source/WebCore/loader/LocalFrameLoaderClient.h:
* Source/WebCore/loader/ResourceMonitor.cpp:
(WebCore::ResourceMonitor::checkNetworkUsageExcessIfNecessary):
* Source/WebCore/loader/ResourceMonitorThrottler.cpp: Added.
(WebCore::ResourceMonitorThrottler::ResourceMonitorThrottler):
(WebCore::ResourceMonitorThrottler::throttlerForHost):
(WebCore::ResourceMonitorThrottler::removeOldestThrottler):
(WebCore::ResourceMonitorThrottler::tryAccess):
(WebCore::ResourceMonitorThrottler::AccessThrottler::tryAccessAndUpdateHistory):
(WebCore::ResourceMonitorThrottler::AccessThrottler::oldestAccessTime const):
(WebCore::ResourceMonitorThrottler::AccessThrottler::tryExpire):
* Source/WebCore/loader/ResourceMonitorThrottler.h: Added.
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::showResourceMonitoringError):
(WebCore::LocalFrame::networkUsageDidExceedThreshold): Deleted.
* Source/WebCore/page/LocalFrame.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::shouldOffloadIFrameForHost):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
(WebKit::WebsiteDataStore::resourceMonitorThrottler):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::didExceedNetworkUsageThreshold):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.h:
* Tools/TestWebKitAPI/CMakeLists.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/ResourceMonitor.cpp: Added.
(TestWebKitAPI::ResouceMonitorTest::now):
(TestWebKitAPI::ResouceMonitorTest::later):
(TestWebKitAPI::TEST_F(ResouceMonitorTest, ThrottlerBasic)):
(TestWebKitAPI::TEST_F(ResouceMonitorTest, ThrottlerMaxHosts)):
(TestWebKitAPI::TEST_F(ResouceMonitorTest, ThrottlerLeastRecentAccessedHostWillBeRemoved)):
(TestWebKitAPI::TEST_F(ResouceMonitorTest, ThrottlerEmptyHostname)):

Canonical link: <a href="https://commits.webkit.org/288427@main">https://commits.webkit.org/288427@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c6f1d6dd1bab34394f0afafa3e901b41034a9c59

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/83216 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/2841 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/37504 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/88306 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/34239 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/85308 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2922 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/10776 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/64763 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/34239 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/86270 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/2129 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75643 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/45047 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2033 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29841 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/33283 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/73176 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/30557 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/89676 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/10489 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/7549 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/73198 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/10716 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71460 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/72427 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17945 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16620 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/15351 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/1832 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/10443 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/15924 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/10301 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/13771 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/12072 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.cpp:
```diff
@@ -0,0 +1,146 @@
+    if (it != m_throttlersByHost.end())
+        return it->value;
+
+    auto result = m_throttlersByHost.add(host.toString(), AccessThrottler { });
```
We avoid double hash map lookup in WebKit for performance reasons, this is better rewritten as:
```
return m_throttlersByHost.ensure<StringViewHashTranslator>(host, [] {
    return AccessThrottler { };
}).iterator->value;
```

without the `find` call above.

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.cpp:
```diff
@@ -0,0 +1,146 @@
+    RESOURCEMONITOR_RELEASE_LOG("Initialized with count: %zu, duration: %.f, maxHosts: %zu", count, duration.value(), maxHosts);
+}
+
+auto ResourceMonitorThrottler::throttlerForHost(StringView&& host) -> AccessThrottler&
```
No need for `&&` for StringView.

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.cpp:
```diff
@@ -0,0 +1,146 @@
+    m_throttlersByHost.remove(oldestKey);
+}
+
+bool ResourceMonitorThrottler::tryAccess(StringView&& host)
```
No need for `&&` for StringView.

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.cpp:
```diff
@@ -0,0 +1,146 @@
+
+bool ResourceMonitorThrottler::tryAccess(StringView&& host)
+{
+    return tryAccess(WTFMove(host), ApproximateTime::now());
```
No `WTFMove()`

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.cpp:
```diff
@@ -0,0 +1,146 @@
+    return tryAccess(WTFMove(host), ApproximateTime::now());
+}
+
+bool ResourceMonitorThrottler::tryAccess(StringView&& host, ApproximateTime time)
```
No need for `&&` for StringView.

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.cpp:
```diff
@@ -0,0 +1,146 @@
+    if (host.isEmpty())
+        return false;
+
+    auto& throttler = throttlerForHost(WTFMove(host));
```
No need for `WTFMove()`

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.h:
```diff
@@ -0,0 +1,78 @@
+    WEBCORE_EXPORT ResourceMonitorThrottler();
+    WEBCORE_EXPORT ResourceMonitorThrottler(size_t count, Seconds duration, size_t maxHosts);
+
+    WEBCORE_EXPORT bool tryAccess(StringView&& host);
```
No need for `&&` with StringView

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.h:
```diff
@@ -0,0 +1,78 @@
+    WEBCORE_EXPORT ResourceMonitorThrottler(size_t count, Seconds duration, size_t maxHosts);
+
+    WEBCORE_EXPORT bool tryAccess(StringView&& host);
+    WEBCORE_EXPORT bool tryAccess(StringView&& host, ApproximateTime);
```
No need for `&&` with StringView

bnham commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -15857,6 +15858,17 @@ void WebPageProxy::setPresentingApplicationAuditToken(const audit_token_t& prese
 #endif
 
+#if ENABLE(CONTENT_EXTENSIONS)
+void WebPageProxy::didExeedFrameNetworkUsage(WebCore::FrameIdentifier frameID, String host)
```
nit: `Exceed` not `Exeed` here and throughout the patch.

cdumez replied:
`const String&`

bnham commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -15857,6 +15858,17 @@ void WebPageProxy::setPresentingApplicationAuditToken(const audit_token_t& prese
+
+    auto granted = throttler.tryAccess(WTFMove(host));
+    if (granted)
+        send(Messages::WebPage::DidExeedFrameNetworkUsage(frameID));
```
Is there a reason this is sending a message rather than defining this IPC as one that returns a value?

basuke replied:
Changed to use sendWithAsyncReply.

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
```diff
@@ -2022,6 +2026,27 @@ bool WebLocalFrameLoaderClient::siteIsolationEnabled() const
+    if (url.isEmpty())
+        return;
+
+    webPage->send(Messages::WebPageProxy::DidExeedFrameNetworkUsage(m_frame->frameID(), url.host().toString()), 0);
```
Typo: "Exceed"

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -3286,6 +3286,11 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
 
+
+#if ENABLE(CONTENT_EXTENSIONS)
+    void didExeedFrameNetworkUsage(WebCore::FrameIdentifier, String host);
```
`const String&`

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -15857,6 +15858,17 @@ void WebPageProxy::setPresentingApplicationAuditToken(const audit_token_t& prese
+{
+    auto& throttler = WebCore::ResourceMonitorManager::singleton().throttler();
+
+    auto granted = throttler.tryAccess(WTFMove(host));
```
No need for `WTFMove()`

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -15857,6 +15858,17 @@ void WebPageProxy::setPresentingApplicationAuditToken(const audit_token_t& prese
+
+    auto granted = throttler.tryAccess(WTFMove(host));
+    if (granted)
+        send(Messages::WebPage::DidExeedFrameNetworkUsage(frameID));
```
Typo: `Exceed`

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -15857,6 +15858,17 @@ void WebPageProxy::setPresentingApplicationAuditToken(const audit_token_t& prese
+#if ENABLE(CONTENT_EXTENSIONS)
+void WebPageProxy::didExeedFrameNetworkUsage(WebCore::FrameIdentifier frameID, String host)
+{
+    auto& throttler = WebCore::ResourceMonitorManager::singleton().throttler();
```
Using a singleton here is fishy, odds are the object should be owned by the WebsiteDataStore. It seems risky to use the same data structure for regular browsing and private browsing for example. Similarly, when closing a private browsing tab, it seems wrong to leave data such as host names from the private browsing session in this shared data structure.

cdumez commented on Source/WebCore/loader/ResourceMonitor.cpp:
```diff
@@ -128,7 +128,12 @@ void ResourceMonitor::checkNetworkUsageExcessIfNecessary()
-        protectedFrame()->networkUsageDidExceedThreshold();
+
+        // If the frame has sticky user activation, don't do offloading.
+        if (RefPtr protectedWindow = protectedFrame()->window(); protectedWindow && protectedWindow->hasStickyActivation())
```
Please store the frame in a local variable to avoid repeated calls to `protectedFrame()`

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.h:
```diff
@@ -0,0 +1,78 @@
+        bool tryAccessAndUpdateHistory(ApproximateTime, const Config&);
+        bool tryExpire(ApproximateTime, const Config&);
+        ApproximateTime oldestAccessTime() const;
+        ApproximateTime recentAccessTime() const { return m_recentAccessTime; }
```
`recentAccessTime` -> `newestAccessTime`

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.h:
```diff
@@ -0,0 +1,78 @@
+        void removeExpired(ApproximateTime);
+
+        PriorityQueue<ApproximateTime> m_accessTimes;
+        ApproximateTime m_recentAccessTime { -ApproximateTime::infinity() };
```
`m_newestAccessTime`

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.h:
```diff
@@ -0,0 +1,78 @@
+    void removeExpiredThrottler();
+    void removeOldestThrottler();
+
+    Config m_config { };
```
No need for `{ }`

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.h:
```diff
@@ -0,0 +1,78 @@
+        ApproximateTime m_recentAccessTime { -ApproximateTime::infinity() };
+    };
+
+    auto throttlerForHost(StringView&& host) -> AccessThrottler&;
```
No need for `&&` with `StringView`. We just pass `StringView` by value.

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.cpp:
```diff
@@ -0,0 +1,146 @@
+
+ApproximateTime ResourceMonitorThrottler::AccessThrottler::oldestAccessTime() const
+{
+    ASSERT(!m_accessTimes.isEmpty());
```
Not needed, `peek()` will already do such as check, on release builds even.

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.cpp:
```diff
@@ -0,0 +1,146 @@
+            return it.value.tryExpire(time, m_config);
+        });
+
+        // If there's still too many hosts, then remove the ol
```
You're missing the end of the comment here.

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.h:
```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
```
`2025`?

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -10276,6 +10276,37 @@ void WebPage::setPresentingApplicationAuditToken(CoreIPCAuditToken&& presentingA
+    coreLocalFrame->showResourceMonitoringError(generateResourceMonitorErrorHTML());
+
+    if (RefPtr document = coreLocalFrame->document())
+        document->addConsoleMessage(MessageSource::ContentBlocker, MessageLevel::Error, "Frame was unloaded because its network usage exceeded the limit."_s);
```
Should this be moved to `showResourceMonitoringError()`? We want as much as possible of the implementation to be in WebCore so it can potentially be shared with WebKitLegacy in the future.

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.cpp:
```diff
@@ -0,0 +1,142 @@
+
+auto ResourceMonitorThrottler::throttlerForHost(const String& host) -> AccessThrottler&
+{
+    return m_throttlersByHost.ensure<StringViewHashTranslator>(host, [] {
```
Doesn't make sense to use `<StringViewHashTranslator>` now that `host` is a String instead of a StringView. Just use `ensure()`

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.cpp:
```diff
@@ -0,0 +1,142 @@
+            return it.value.tryExpire(time, m_config);
+        });
+
+        // If there's still too many hosts, then remove oldest one.
```
"If there are still..."

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.cpp:
```diff
@@ -0,0 +1,142 @@
+
+        m_accessTimes.dequeue();
+    }
+    // Tell caller that the queue is empty
```
Comments in WebKit need to end with a period.

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.h:
```diff
@@ -0,0 +1,78 @@
+
+class ResourceMonitorThrottler final {
+    WTF_MAKE_FAST_ALLOCATED;
+    friend MainThreadNeverDestroyed<ResourceMonitorThrottler>;
```
I doubt this is still needed

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.h:
```diff
@@ -0,0 +1,78 @@
+    WEBCORE_EXPORT ResourceMonitorThrottler();
+    WEBCORE_EXPORT ResourceMonitorThrottler(size_t count, Seconds duration, size_t maxHosts);
+
+    WEBCORE_EXPORT bool tryAccess(const String& host);
```
Not needed, see comment below.

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.h:
```diff
@@ -0,0 +1,78 @@
+    WEBCORE_EXPORT ResourceMonitorThrottler(size_t count, Seconds duration, size_t maxHosts);
+
+    WEBCORE_EXPORT bool tryAccess(const String& host);
+    WEBCORE_EXPORT bool tryAccess(const String& host, ApproximateTime);
```
```
WEBCORE_EXPORT bool tryAccess(const String& host, ApproximateTime = ApproximateTime::now());
```

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.h:
```diff
@@ -0,0 +1,78 @@
+        ApproximateTime m_newestAccessTime { -ApproximateTime::infinity() };
+    };
+
+    auto throttlerForHost(const String& host) -> AccessThrottler&;
```
`AccessThrottler& throttlerForHost(const String& host);` ? I don't see the point of the `auto` in this instance.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -15857,6 +15857,14 @@ void WebPageProxy::setPresentingApplicationAuditToken(const audit_token_t& prese
+#if ENABLE(CONTENT_EXTENSIONS)
+void WebPageProxy::shouldOffloadIFrameForHost(const String& host, CompletionHandler<void(bool)>&& completionHandler)
+{
+    auto granted = protectedWebsiteDataStore()->resourceMonitorThrottler().tryAccess(host);
```
`auto` -> `bool`. Also, boolean variables need a prefix in WebKit as per coding style. Here I would call it `wasGranted`.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -3286,6 +3286,11 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
 
+
+#if ENABLE(CONTENT_EXTENSIONS)
+    void shouldOffloadIFrameForHost(const String& host, CompletionHandler<void(bool)>&&);
```
Can this be `const`?

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
```diff
@@ -2022,6 +2026,30 @@ bool WebLocalFrameLoaderClient::siteIsolationEnabled() const
+    if (url.isEmpty())
+        return;
+
+    webPage->sendWithAsyncReply(Messages::WebPageProxy::ShouldOffloadIFrameForHost(url.host().toString()), [frame = m_frame->coreLocalFrame()] (bool granted) {
```
`wasGranted`

cdumez replied:
`toString()` -> `toStringWithoutCopying()`

cdumez commented on Tools/TestWebKitAPI/Tests/WebCore/ResourceMonitor.cpp:
```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
```
`2025`

cdumez reviewed (approved):
LGTM with fixes.

basuke committed:
Add throttler to limit the ResourceMonitor activity.
https://bugs.webkit.org/show_bug.cgi?id=285107
rdar://141840417

Reviewed by Chris Dumez.

When ResourceMonitor detects large resource usage in a iframe, currently the decision is in
action immediately. In this patch, add throttler to limit the execution of the decision.

- Moving the action from LocalFrame to LocalFrameLoaderClient.
- Ask WebPageProxy for the decision.
- Add ResourceMonitorThrottler to limit the action.
- Place the throttler in ResourceMonitorManager for the global decision.
- Send the result to WebPage if the decision is ongoing.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/loader/LocalFrameLoaderClient.cpp:
(WebCore::LocalFrameLoaderClient::didExceedNetworkUsageThreshold):
* Source/WebCore/loader/LocalFrameLoaderClient.h:
* Source/WebCore/loader/ResourceMonitor.cpp:
(WebCore::ResourceMonitor::checkNetworkUsageExcessIfNecessary):
* Source/WebCore/loader/ResourceMonitorThrottler.cpp: Added.
(WebCore::ResourceMonitorThrottler::ResourceMonitorThrottler):
(WebCore::ResourceMonitorThrottler::throttlerForHost):
(WebCore::ResourceMonitorThrottler::removeOldestThrottler):
(WebCore::ResourceMonitorThrottler::tryAccess):
(WebCore::ResourceMonitorThrottler::AccessThrottler::tryAccessAndUpdateHistory):
(WebCore::ResourceMonitorThrottler::AccessThrottler::oldestAccessTime const):
(WebCore::ResourceMonitorThrottler::AccessThrottler::tryExpire):
* Source/WebCore/loader/ResourceMonitorThrottler.h: Added.
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::showResourceMonitoringError):
(WebCore::LocalFrame::networkUsageDidExceedThreshold): Deleted.
* Source/WebCore/page/LocalFrame.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::shouldOffloadIFrameForHost):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
(WebKit::WebsiteDataStore::resourceMonitorThrottler):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::didExceedNetworkUsageThreshold):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.h:
* Tools/TestWebKitAPI/CMakeLists.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/ResourceMonitor.cpp: Added.
(TestWebKitAPI::ResouceMonitorTest::now):
(TestWebKitAPI::ResouceMonitorTest::later):
(TestWebKitAPI::TEST_F(ResouceMonitorTest, ThrottlerBasic)):
(TestWebKitAPI::TEST_F(ResouceMonitorTest, ThrottlerMaxHosts)):
(TestWebKitAPI::TEST_F(ResouceMonitorTest, ThrottlerLeastRecentAccessedHostWillBeRemoved)):
(TestWebKitAPI::TEST_F(ResouceMonitorTest, ThrottlerEmptyHostname)):

Canonical link: https://commits.webkit.org/288427@main
